### PR TITLE
Initial draft of hardware requirement helpers

### PR DESCRIPTION
**Pull request by happz**
Adapted from implementation by Artemis (https://gitlab.com/testing-farm/artemis)

Related to #1314.

### CODE REVIEW

**Comment by happz** [comment]
A few commands for experimenting:

```shell
tmt -vv run plan --default provision -h beaker --hardware "cpu.processors>=4" --hardware "memory>=4GiB" finish

# What happens when a HW requirement is not supported by plugin yet? Not much for now:
tmt -vv run plan --default provision -h beaker --hardware "cpu.processors>=4" --hardware "memory>=4GiB" --hardware "cpu.family-name=Haswell" finish

# testcloud will need more work to amend its own options --memory/--disk and --hardware - for now, --hardware is ignored:
$ tmt -vv run plan --default provision -h virtual --hardware "cpu.processors>=4" --hardware "memory>=4GiB" finish
```

**Comment by martinhoyer** [comment]
Just fyi, pint project is now Python 3.9+ only.
EPEL9 runs 0.16.1 (2020-09-22), though theoretically last Python 3.8 compatible upstream version is 0.22 (2023-05-25)
I found an el8 0.10.1 (2020-01-07) rpm, but don't actually see it in any repo. Is this a draft for next year's c8s eol? :)

**Comment by happz** [comment]
> Just fyi, pint project is now Python 3.9+ only. EPEL9 runs 0.16.1 (2020-09-22), though theoretically last Python 3.8 compatible upstream version is 0.22 (2023-05-25) I found an el8 0.10.1 (2020-01-07) rpm, but don't actually see it in any repo. Is this a draft for next year's c8s eol? :)

Chm, this Python 3.6 compatibility is going to kill me... This or high levels of blood sugar, I don't know which is worse :/

I think I tried to freeze the requirement to some 0.18.something version of Pint, mostly because of type annotations, and we discussed bumping the epel versions to the one in these requirements with @lukaszachy, but that's still WiP.

File: tmt/hardware.py [hunk 1]
```diff
@@ -0,0 +1,1088 @@
...
+            self.__class__.__name__.lower(): [
+                constraint.to_spec() for constraint in self.constraints
+                ]
+            }
+
+    def uses_constraint(self, parent: tmt.utils.Common, constraint_name: str) -> bool:
+        """
+        Inspect constraint whether it or its children use a constraint of a given name.
+
+        :param logger: logger to use for logging.
```
Comment (psss) [inline tmt/hardware.py:370]: Another `parent` versus `logger`.

File: tmt/hardware.py [hunk 2]
```diff
@@ -0,0 +1,1088 @@
...
+            members: Optional[List[ConstraintBase]] = None
+            ) -> Iterator[List[ConstraintBase]]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        Since the ``and`` reducer demands all child constraints must be satisfied, and some of
+        these constraints can also be compound constraints, we need to construct a cartesian
+        product of spans yielded by child constraints to include all possible combinations.
+
+        :param logger: logger to use for logging.
```
Comment (psss) [inline tmt/hardware.py:395]: Not provided.

File: tmt/hardware.py [hunk 3]
```diff
@@ -0,0 +1,1088 @@
...
+    """
+    A constraint imposing a particular limit to one of the guest properties.
+    """
+
+    # Name of the constraint. Used for logging purposes, usually matches the
+    # name of the system property.
+    name: str
+
+    # A binary operation to use for comparing the constraint value and the
+    # value specified by system or flavor.
```
Comment (psss) [inline tmt/hardware.py:416]: The `system or flavor` are a bit irrelevant here?

File: tmt/hardware.py [hunk 4]
```diff
@@ -0,0 +1,1088 @@
...
+        self.constraint_name = constraint_name
+        self.raw_value = raw_value
+
+
+#
+# Constraint classes
+#
+
+@dataclasses.dataclass(repr=False)
+class ConstraintBase(tmt.utils.SpecBasedContainer):
```
Comment (psss) [inline tmt/hardware.py:275]: We have `CompoundContraint` defined below, what about calling this `BaseConstraint` to keep consistency?

File: tmt/hardware.py [hunk 5]
```diff
@@ -0,0 +1,1088 @@
...
+        """
+
+        self.operator = operator
+        self.operator_handler = OPERATOR_TO_HANDLER[operator]
+
+    def uses_constraint(self, parent: tmt.utils.Common, constraint_name: str) -> bool:
+        """
+        Inspect constraint whether it or its children use a constraint of a given name.
+
+        :param logger: logger to use for logging.
```
Comment (psss) [inline tmt/hardware.py:548]: `logger` instead of `parent`.

File: tmt/hardware.py [hunk 6]
```diff
@@ -0,0 +1,1088 @@
...
+            members: Optional[List[ConstraintBase]] = None
+            ) -> Iterator[List[ConstraintBase]]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        Since the ``and`` reducer demands all child constraints must be satisfied, and some of
+        these constraints can also be compound constraints, we need to construct a cartesian
+        product of spans yielded by child constraints to include all possible combinations.
+
+        :param logger: logger to use for logging.
```
Comment (psss) [inline tmt/hardware.py:582]: No `logger` in method definition.

File: tmt/hardware.py [hunk 7]
```diff
@@ -0,0 +1,1088 @@
...
+            members: Optional[List[ConstraintBase]] = None
+            ) -> Iterator[List[ConstraintBase]]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        Since the ``any`` reducer allows any child constraints to be satisfied for the whole group
+        to evaluate as ``True``, it is trivial to generate spans - each child constraint shall
+        provide its own "branch", and there is no need for products or joins of any kind.
+
+        :param logger: logger to use for logging.
```
Comment (psss) [inline tmt/hardware.py:638]: No `logger` in the method definition.

File: tmt/hardware.py [hunk 8]
```diff
@@ -0,0 +1,1088 @@
...
+        for constraint in self.constraints:
+            for span in constraint.spans():
+                yield members + span
+
+
+#
+# Constraint parsing
+#
+
+def ungroupify(fn: Callable[[SpecType], ConstraintBase]) -> Callable[[SpecType], ConstraintBase]:
```
Comment (psss) [inline tmt/hardware.py:655]: Please, add a docstring. Similarly below.

File: tmt/hardware.py [hunk 9]
```diff
@@ -0,0 +1,1088 @@
...
+
+    for distro in spec.get('distro', []):
+        constraint = Constraint.from_specification('compatible.distro', distro, as_quantity=False)
+
+        constraint.change_operator(Operator.CONTAINS)
+
+        group.constraints.append(constraint)
+
+    if len(group.constraints) == 1:
+        return group.constraints[0]
```
Comment (psss) [inline tmt/hardware.py:771]: Isn't this covered by `ungroupify` already?

File: tmt/hardware.py [hunk 10]
```diff
@@ -0,0 +1,1088 @@
...
+    :returns: block representation as :py:class:`ConstraintBase` or one of its subclasses.
+    """
+
+    group = And()
+
+    group.constraints += [
+        Constraint.from_specification(
+            f'disk[{disk_index}].{constraint_name}', str(
+                spec[constraint_name])) for constraint_name in (
+            'size',) if constraint_name in spec]
```
Comment (psss) [inline tmt/hardware.py:837]: Perhaps wrap `for` and `if` to new lines as done elsewhere?

File: tmt/hardware.py [hunk 11]
```diff
@@ -0,0 +1,1088 @@
...
+
+    if 'version' in spec:
+        group.constraints.append(
+            Constraint.from_specification(
+                'tpm.version',
+                spec['version'],
+                as_quantity=False))
+
+    if len(group.constraints) == 1:
+        return group.constraints[0]
```
Comment (psss) [inline tmt/hardware.py:928]: Covered by `ungroupify`?

File: tmt/hardware.py [hunk 12]
```diff
@@ -0,0 +1,1088 @@
...
+            spec=spec
+            )
+
+    def to_spec(self) -> Dict[str, Any]:
+        return self.spec  # type: ignore[no-any-return]
+
+    def to_minimal_spec(self) -> Dict[str, Any]:
+        return self.spec  # type: ignore[no-any-return]
+
+    def show_support(self, check: Callable[[Constraint], bool], logger: tmt.log.Logger) -> None:
```
Comment (psss) [inline tmt/hardware.py:1075]: Please, add a docstring.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks much for preparing this! Looks very good. Added a bunch of comments, but mostly just minor suggestions.

File: tmt/steps/provision/connect.py [hunk 13]
```diff
@@ -103,6 +103,9 @@ def go(self) -> None:
         else:
             self.debug('Using private key authentication.')
 
+        if data.hardware and data.hardware.constraint:
+            self.warn("'connect' provision plugin does not support hardware requirements")
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/steps/provision/connect.py]: Fixed in e7c0774a.

File: tmt/steps/provision/mrack.py [hunk 14]
```diff
@@ -137,51 +162,55 @@ def _parse_amount(self, in_string: str) -> Tuple[str, int]:
...
-            for key, val in hw.items():
-                if key == "memory":
-                    operator, amount = self._parse_amount(val)
-                    system.update({
-                        key: {
-                            value: amount,
-                            op: operator
+            for constraint in hw.constraint.span():
+                if not isinstance(constraint, tmt.hardware.Constraint):
+                    raise tmt.utils.ProvisionError(f'unsupported hw constraint {constraint}')
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/steps/provision/mrack.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 15]
```diff
@@ -0,0 +1,1088 @@
+"""
+Guest hardware requirements specification and helpers.
+
+TMT allows metadata describe various HW requirements a guest needs to satisfy.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 16]
```diff
@@ -0,0 +1,1088 @@
...
+
+    :param text: string to examine.
+    :param pattern: regular expression.
+    :returns: ``True`` if pattern matches the string.
+    """
+
+    return re.match(pattern, text) is not None
+
+
+def notmatch(text: str, pattern: str) -> bool:
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 17]
```diff
@@ -0,0 +1,1088 @@
...
+class ConstraintBase(tmt.utils.SpecBasedContainer):
+    """
+    Base class for all classes representing one or more constraints.
+    """
+
+    @classmethod
+    def from_spec(cls, spec: Any) -> 'ConstraintBase':
+        return parse_hw_requirements(spec)
+
+    def to_(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 18]
```diff
@@ -0,0 +1,1088 @@
...
+
+    def to_minimal_spec(self) -> Dict[str, Any]:
+        return self.to_spec()
+
+    def uses_constraint(self, parent: tmt.utils.Common, constraint_name: str) -> bool:
+        """
+        Inspect constraint whether the constraint or one of its children use a constraint of
+        a given name.
+
+        :param logger: logger to use for logging.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 19]
```diff
@@ -0,0 +1,1088 @@
...
+            members: Optional[List['ConstraintBase']] = None
+            ) -> Iterator[List['ConstraintBase']]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        For a trivial constraint, there is only one span, and that's the constraint itself. In the
+        case of compound constraints, the set of spans would be bigger, depending on the
+        constraint's ``reducer``.
+
+        :param logger: logger to use for logging.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 1 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 2 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 4 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 5 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 6 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 7 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 9 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 10 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py (hunk 11 above)
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/steps/provision/__init__.py [hunk 20]
```diff
@@ -330,6 +333,34 @@ def sync(self, guest: 'Guest') -> None:
         self.in_sync = True
 
 
+def normalize_hardware(
+        key_address: str,
+        raw_hardware: Optional[tmt.hardware.SpecType],
+        logger: tmt.log.Logger) -> Optional[tmt.hardware.Hardware]:
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/steps/provision/__init__.py:339]: Added in e7c0774a.

File: tmt/hardware.py [hunk 21]
```diff
@@ -0,0 +1,1088 @@
...
+
+# Regular expression to match and split the `value`` part of the key:value pair.
+# Said `value` bundles together the operator, the actual value of the constraint,
+# and units.
+VALUE_PATTERN = re.compile(
+    r'^(?P<operator>==|!=|~|!~|=~|=|>=|>|<=|<|contains|(?:not contains))?\s*(?P<value>.+?)\s*$')
+
+# TODO: provide comment
+PROPERTY_PATTERN = re.compile(r'(?P<property_name>[a-z_]+)(?:\[(?P<index>[+-]?\d+)\])?')
+# TODO: provide comment
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Added in e7c0774a.

File: tmt/hardware.py [hunk 22]
```diff
@@ -0,0 +1,1088 @@
...
+    """
+    Find out whether an item is in the given list.
+
+    .. note::
+
+       Opposite of :py:func:`operator.contains`.
+
+    :param haystack: container to examine.
+    :param needle: item to look for in ``haystack``.
+    :returns: ``True`` if ``needle`` is in ``haystack``.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Fixed in e7c0774a.

File: tmt/hardware.py [hunk 23]
```diff
@@ -0,0 +1,1088 @@
...
+
+        :param logger: logger to use for logging.
+        :param members: if specified, each span generated by this method is prepended with this
+            list.
+        :yields: iterator over all spans.
+        """
+
+        yield (members or []) + [self]
+
+    def span(self) -> List['ConstraintBase']:
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Added in e7c0774a.

File: tmt/hardware.py (hunk 8 above)
Comment (happz) [reply tmt/hardware.py]: Added in e7c0774a.

File: tmt/hardware.py (hunk 12 above)
Comment (happz) [reply tmt/hardware.py:1150]: Added in e7c0774a.

File: tmt/hardware.py [hunk 24]
```diff
@@ -0,0 +1,1088 @@
...
+    NOTMATCH = '!~'
+    CONTAINS = 'contains'
+    NOTCONTAINS = 'not contains'
+
+
+# Regular expression to match and split the `value`` part of the key:value pair.
+# Said `value` bundles together the operator, the actual value of the constraint,
+# and units.
+VALUE_PATTERN = re.compile(
+    r'^(?P<operator>==|!=|~|!~|=~|=|>=|>|<=|<|contains|(?:not contains))?\s*(?P<value>.+?)\s*$')
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py]: Probably not, but both `contains` and `not contains` have no place in this particular pattern. Refactored in e7c0774a.

File: tmt/hardware.py (hunk 3 above)
Comment (happz) [reply tmt/hardware.py]: Actually, I rephrased the sentence, and the meaning is still valid. The only user so far, in tmt repo, is the beaker plugin, which does no comparison but rewrites constraints into XML and leaves comparisons and filtering to Beaker. But for `virtual` or `podman`, the comparison will happen some day.

Comment (psss) [reply tmt/hardware.py]: Thanks for clarification.

File: tmt/steps/provision/mrack.py [hunk 25]
```diff
@@ -104,84 +133,55 @@ def import_and_load_mrack_deps(workdir: Any, name: str) -> None:
...
-                break
+            system = {}
+            disks = []
+            cpu = {}
 
-            assert len(result) == 1
-            assert len(amount) >= 1
+            for constraint in hw.constraint.variant():
+                if not isinstance(constraint, tmt.hardware.Constraint):
+                    raise tmt.utils.ProvisionError(f"Unsupported hw constraint '{constraint}'")
```
Comment (psss) [inline tmt/steps/provision/mrack.py:147]:
```suggestion
                    raise tmt.utils.ProvisionError(f"Unsupported hardware constraint '{constraint}'.")
```

File: tmt/hardware.py [hunk 26]
```diff
@@ -0,0 +1,1176 @@
...
+            spec=spec
+            )
+
+    def to_spec(self) -> Dict[str, Any]:
+        return self.spec  # type: ignore[no-any-return]
+
+    def to_minimal_spec(self) -> Dict[str, Any]:
+        return self.spec  # type: ignore[no-any-return]
+
+    def show_support(self, check: Callable[[Constraint], bool], logger: tmt.log.Logger) -> None:
```
Comment (psss) [inline tmt/hardware.py:1150]:
```suggestion
    def check_support(self, check: Callable[[Constraint], bool], logger: tmt.log.Logger) -> None:
```

Wouldn't `check` be a bit more appropriate as the main goal of the method seems to be verifying the constraints?

**Review by psss — CHANGES_REQUESTED** [review]
Just two nitpicks, otherwise looks good.

File: tmt/hardware.py (hunk 26 above)
Comment (psss) [reply tmt/hardware.py:1150]: Ok, understood. What about `examine_support()` or `explore_support()`?

Comment (happz) [reply tmt/hardware.py]: Renamed to `report_support()` in dc5b7deb394fcbec89dd5bc4197ad4c566022076

File: tmt/steps/provision/mrack.py (hunk 25 above)
Comment (happz) [reply tmt/steps/provision/mrack.py]: Fixed in dc5b7deb394fcbec89dd5bc4197ad4c566022076

File: tmt/steps/provision/mrack.py [hunk 27]
```diff
@@ -40,26 +42,204 @@
...
+    # and that may be weird :/
+    return 'like', value.replace('.*', '%').replace('.+', '%')
+
+
+# Transcription of our HW constraints into Mrack's own representation. It's based
+# on dictionaries, and it's slightly weird. There is no distinction between elements
+# that do not have attributes, like <and/>, and elements that must have them, like
+# <memory/> and other binary operations. Also, there is no distinction betwen
+# element attribute and child element, both are specified as dictionary key, just
+# the former would be a string, the latter another, nested, disctionary.
```
Comment (psss) [inline tmt/steps/provision/mrack.py:84]:
```suggestion
# the former would be a string, the latter another, nested, dictionary.
```

File: tmt/steps/provision/mrack.py [hunk 28]
```diff
@@ -100,108 +280,30 @@ def import_and_load_mrack_deps(workdir: Any, name: str) -> None:
...
+                    for constraint in hw.constraint.variant()
+                    ])
 
-                result.append(str(multiplier * int(amount.split(size, maxsplit=1)[0])))
-                break
+            logger.info('Transformed hardware', tmt.utils.dict_to_yaml(transformed.to_mrack()))
 
-            # returns operator, amount
-            return result[0], int(result[1])
+            print(tmt.utils.dict_to_yaml(transformed.to_mrack()))
```
Comment (psss) [inline tmt/steps/provision/mrack.py:296]: A debugging leftover?

**Review by psss — CHANGES_REQUESTED** [review]
Looks good. Just two nitpicks found.

File: tmt/steps/provision/mrack.py (hunk 27 above)
Comment (happz) [reply tmt/steps/provision/mrack.py]: Fixed in 67aac676

File: tmt/steps/provision/mrack.py (hunk 28 above)
Comment (happz) [reply tmt/steps/provision/mrack.py]: Fixed in 67aac676

**Review by psss — APPROVED** [review]

**Pull request merged**